quantum/split_common/transport.c

```diff
@@ -3,187 +3,124 @@
-  err = i2c_master_write(I2C_KEYMAP_START);
-  if (err) { goto i2c_error; }
-
-  // Start read
-  err = i2c_master_start(SLAVE_I2C_ADDRESS + I2C_READ);
-  if (err) { goto i2c_error; }
-
-  if (!err) {
-    int i;
-    for (i = 0; i < ROWS_PER_HAND-1; ++i) {
-      matrix[i] = i2c_master_read(I2C_ACK);
-    }
-    matrix[i] = i2c_master_read(I2C_NACK);
-    i2c_master_stop();
-  } else {
-i2c_error: // the cable is disconnceted, or something else went wrong
-    i2c_reset_state();
-    return false;
+#  ifdef BACKLIGHT_ENABLE
+  static uint8_t prev_level = ~0;
+  uint8_t        level      = get_backlight_level();
+  if (level != prev_level) {
+    i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_BACKLIT_START, (void *)&level, sizeof(level), TIMEOUT);
+    prev_level = level;
   }
-
-#ifdef RGBLIGHT_ENABLE
-  if (RGB_DIRTY) {
-    err = i2c_master_start(SLAVE_I2C_ADDRESS + I2C_WRITE);
-    if (err) { goto i2c_error; }
-
-    // RGB Location
-    err = i2c_master_write(I2C_RGB_START);
-    if (err) { goto i2c_error; }
-
-    uint32_t dword = eeconfig_read_rgblight();
-
-    // Write RGB
-    err = i2c_master_write_data(&dword, 4);
-    if (err) { goto i2c_error; }
-
-    RGB_DIRTY = false;
-    i2c_master_stop();
+#  endif
+
+#  ifdef RGBLIGHT_ENABLE
+  static uint32_t prev_rgb = ~0;
+  uint32_t        rgb      = eeconfig_read_rgblight();
+  if (rgb != prev_rgb) {
+    i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_BACKLIT_START, (void *)&rgb, sizeof(rgb), TIMEOUT);
+    prev_rgb = rgb;
   }
-#endif
+#  endif
 
   return true;
 }
 
 void transport_slave(matrix_row_t matrix[]) {
-
-  for (int i = 0; i < ROWS_PER_HAND; ++i)
-  {
-    i2c_slave_buffer[I2C_KEYMAP_START + i] = matrix[i];
-  }
-  // Read Backlight Info
-  #ifdef BACKLIGHT_ENABLE
-  if (BACKLIT_DIRTY)
-  {
-    backlight_set(i2c_slave_buffer[I2C_BACKLIT_START]);
-    BACKLIT_DIRTY = false;
+  for (int i = 0; i < ROWS_PER_HAND * sizeof(matrix_row_t); ++i) {
+    i2c_slave_reg[I2C_KEYMAP_START + i] = matrix[i];
   }
-  #endif
-  #ifdef RGBLIGHT_ENABLE
-  if (RGB_DIRTY)
-  {
-    // Disable interupts (RGB data is big)
-    cli();
-    // Create new DWORD for RGB data
-    uint32_t dword;
-
-    // Fill the new DWORD with the data that was sent over
-    uint8_t * dword_dat = (uint8_t *)(&dword);
-    for (int i = 0; i < 4; i++)
-    {
-      dword_dat[i] = i2c_slave_buffer[I2C_RGB_START + i];
-    }
-
-    // Update the RGB now with the new data and set RGB_DIRTY to false
-    rgblight_update_dword(dword);
-    RGB_DIRTY = false;
-    // Re-enable interupts now that RGB is set
-    sei();
-  }
-  #endif
-}
 
-void transport_master_init(void) {
-  i2c_master_init();
-}
+// Read Backlight Info
+#  ifdef BACKLIGHT_ENABLE
+  backlight_set(i2c_slave_reg[I2C_BACKLIT_START]);
+#  endif
 
-void transport_slave_init(void) {
-  i2c_slave_init(SLAVE_I2C_ADDRESS);
+#  ifdef RGBLIGHT_ENABLE
+  uint32_t rgb = *(uint32_t *)(i2c_slave_reg + I2C_RGB_START);
+  // Update the RGB with the new data
+  rgblight_update_dword(rgb);
```

Comment:
Every transaction seems to be running rgblight_update_dword (). rgblight_update_dword () is writing eprom, is this OK? @drashna

Reply:
Good cache. 

No, that's not really okay. This will cause major issues for the eeprom in the long term, since it's writing to it, every cycle.

Reply:
It may be worth modifying `rgblight_set` in some way, so that it adds a callback for the split keyboard codie.

Reply:
I opened the #4984.
It is very simple change, I think it will be working, so I did not tested.

Reply:
Merged, and that change should be perfect to help out here.

And I tested eeprom write change already, and it appears to work as expected.